### ISSUE TITLE
Add ReorderableListView.seperated constructor

### ISSUE THREAD

**Issue by phils-hub** [flutter/flutter#76706]
## Steps to Reproduce

1. Run `flutter create bug`.
2. Update the files as follows:
Add a ReorderableListView child to the Scaffold body's child column. The issue is with every ReorderableListView, but the following makes it easy to see, as follows:
````dart
Expanded(
  ReorderableListView(
    children: List.generate(10, (index) => Container(
      key: ValueKey('abcd-$index'),
      width: 250,
      height: 150,
      color: Colors.lightBlue,
      margin: EdgeInsets.symmetric(vertical: 20, horizontal: 25),
    )),
    onReorder: (oldIndex, newIndex) {},
  )
)
`````
3. Long press on any of the children of the ReorderableListView and you will see that a box shadow is applied to the border around the margin of the child that was long pressed. 

**Expected results:** The shadow or animation shown by the ReorderableListView surrounds the actual child without including the margin of the child.

**Actual results:** The shadow or animation shown includes the margin of the child. 

Since it is not possible to define padding or separators for the ReorderableListView's children (unlike with the ListView.separated variant), I would have expected the margin to be respected by the ReorderableListView. Currently i do not know how to avoid this problem.

Screenshot showing the shadow around the margin:
<img src="https://user-images.githubusercontent.com/25553592/109020765-8c040380-76ba-11eb-8b6c-631e5172a846.png" width="300" />

**Comment by HansMuller** [comment]
CC @darrenaustin

**Comment by pedromassangocode** [comment]
<details>
<summary>flutter doctor -v</summary>

```bash
[✓] Flutter (Channel master, 1.27.0-9.0.pre.85, on macOS 11.2.1 20D74 darwin-x64, locale en-AO)
    • Flutter version 1.27.0-9.0.pre.85 at /Users/pedromassango/Code/flutter_master
    • Framework revision 71b2b4d6e0 (2 hours ago), 2021-02-25 00:01:01 -0500
    • Engine revision 867e51b762
    • Dart version 2.13.0 (build 2.13.0-73.0.dev)

[✓] Android toolchain - develop for Android devices (Android SDK version 30.0.2)
    • Android SDK at /Users/pedromassango/Library/Android/sdk
    • Platform android-30, build-tools 30.0.2
    • Java binary at: /Applications/Android Studio.app/Contents/jre/jdk/Contents/Home/bin/java
    • Java version OpenJDK Runtime Environment (build 1.8.0_242-release-1644-b3-6915495)
    • All Android licenses accepted.

[!] Xcode - develop for iOS and macOS
    • Xcode at /Applications/Xcode.app/Contents/Developer
    • Xcode 12.1, Build version 12A7403
    ! CocoaPods 1.9.3 out of date (1.10.0 is recommended).
        CocoaPods is used to retrieve the iOS and macOS platform side's plugin code that responds to your plugin usage on the Dart side.
        Without CocoaPods, plugins will not work on iOS or macOS.
        For more info, see https://flutter.dev/platform-plugins
      To upgrade see https://guides.cocoapods.org/using/getting-started.html#installation for instructions.

[✓] Chrome - develop for the web
    • Chrome at /Applications/Google Chrome.app/Contents/MacOS/Google Chrome

[✓] Android Studio (version 4.1)
    • Android Studio at /Applications/Android Studio.app/Contents
    • Flutter plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/9212-flutter
    • Dart plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/6351-dart
    • Java version OpenJDK Runtime Environment (build 1.8.0_242-release-1644-b3-6915495)

[✓] IntelliJ IDEA Community Edition (version 2020.3.2)
    • IntelliJ at /Applications/IntelliJ IDEA CE.app
    • Flutter plugin version 53.1.5
    • Dart plugin version 203.6912

[✓] VS Code (version 1.52.1)
    • VS Code at /Applications/Visual Studio Code.app/Contents
    • Flutter extension version 3.18.1

[✓] Connected device (2 available)
    • macOS (desktop) • macos  • darwin-x64     • macOS 11.2.1 20D74 darwin-x64
    • Chrome (web)    • chrome • web-javascript • Google Chrome 88.0.4324.192

! Doctor found issues in 1 category.
```

</details>

**Comment by impaulb** [comment]
Seeing the same issue! Wish there was a method similar to ListView.separated for ReorderableListView.

**Comment by darrenaustin** [comment]
Yeah, there is no way the list view can know that the padding or margin you are putting in each item is something that you want to exclude from the dragged item.  The real solution as suggested is to have a ReorderableListView.separated constructor (I updated the issue title accordingly).

I took a quick look at implementing this with just a widget builder wrapper that inserted divider widgets between the children. It sort of works, but causes issues during the item drag, as the usual gap that is shown between items will overlap with where the dividers go.  The internals of the reorderable list will have to know more about the dividers to make this work well.

**Comment by vandir** [comment]
Does anyone know a way to achieve the same results of the `separated` constructor?

**Comment by vannevar-morgan** [comment]
agreed, would be nice to have this added.

**Comment by Toolenaar** [comment]
> Does anyone know a way to achieve the same results of the `separated` constructor?

I fixed this by putting a bottom padding around my ListItemWidget and then use a Stack in the  the proxyDecorator on ReorderableListView  to adjust my "drag UI":

```
proxyDecorator: (child, index, animation) {
       
        return Material(
          elevation: 0,
          color: Colors.transparent,
          child: Stack(
            children: [
              Positioned(
                top: 0,
                left: 0,
                right: 0,
                bottom: 16,
                child: Material(
                  borderRadius: StrackStyles.borderRadius,
                  elevation: 24,
                  color: Colors.black,
                  shadowColor: Colors.black.withOpacity(0.6),
                ),
              ),
              child,
            ],
          ),
        );
      },
```

**Comment by EmanueleVinci** [comment]
any news?

**Comment by itsjasur** [comment]
any update on this?

**Comment by giboin** [comment]
Hi everyone ! I've been using Flutter for a bit less than a year now, I find it amazing (thank you all for your work) and I would like to try and contribute for the first time.

I took a look at this issue and at the implementation of both ReorderableListView and ListView.separated, and I feel like I could use some help about the behavior to expect on the separators when an item is moving.

What seems clear to me is that the separators shouldn't be dragged with the moving item, and that like on the ListView.separated there is no separator before the first item and after the last one.

But should the separators move like other items, like shown on this comment ? https://github.com/flutter/flutter/issues/76706#issuecomment-1067992802
I think it's a little weird that they do this little jump they do to reorder themselves when the user stops dragging though, and also the fact that you can see two separators without an item between them.

Then should separators move like that but merge themselves when they are next to another, and unmerge when needed to separate two unseparated items?

Should one of the separators disappear next to the dragged item, and reappear after the dragging ends?

Or do you have any other idea? Like a way to let the developer somehow change the separators appearance when dragging? 

Thanks for your help

**Comment by SharmaDhruv2511** [comment]
I want to work on this issue. Can you please assign me?

**Comment by AppleDev879** [comment]
I was able to improve a bit on @bstolinski 's solution. For me, the drag handle icons were being drawn for the separators as well as the content cells. I also fixed some of the harsh "snapping" on drag by supplying an animation controller to fade out the separators while drag is in progress. We wait 50 ms to fade them back in so that the cells can be redrawn first.

https://github.com/flutter/flutter/assets/1386393/9347c643-bfc6-49e5-9776-8bbf806b3ae4

```
class CustomReorderableListView extends ReorderableListView {
  CustomReorderableListView.separated({
    Key? key,
    required IndexedWidgetBuilder itemBuilder,
    required IndexedWidgetBuilder separatorBuilder,
    required int itemCount,
    required ReorderCallback onReorder,
    VoidCallback? onReorderStarted,
    Function(int)? onReorderEnded,
    double? itemExtent,
    Widget? prototypeItem,
    ReorderItemProxyDecorator? proxyDecorator,
    bool buildDefaultDragHandles = true,
    EdgeInsets? padding,
    Widget? header,
    Axis scrollDirection = Axis.vertical,
    bool reverse = false,
    ScrollController? scrollController,
    bool? primary,
    ScrollPhysics? physics,
    bool shrinkWrap = false,
    double anchor = 0.0,
    double? cacheExtent,
    DragStartBehavior dragStartBehavior = DragStartBehavior.start,
    ScrollViewKeyboardDismissBehavior keyboardDismissBehavior =
        ScrollViewKeyboardDismissBehavior.manual,
    String? restorationId,
    Clip clipBehavior = Clip.hardEdge,
  }) : super.builder(
          key: key,
          itemCount: max(0, itemCount * 2 - 1),
          itemBuilder: (BuildContext context, int index) {
            final isSeparator = index.isOdd;
            final itemIndex = index ~/ 2;
            final Widget item;

            if (isSeparator) {
              // Separator
              item = separatorBuilder.call(context, itemIndex);
              // Ensure the separator has a unique key
              if (item.key == null) {
                return KeyedSubtree(
                  key: ValueKey('ReorderableSeparator${index}Key'),
                  child: IgnorePointer(child: item),
                );
              }
              // No drag handles for separators
              return item;
            } else {
              // Actual list item
              item = itemBuilder.call(context, itemIndex);
              // Wrap with ReorderableDragStartListener if drag handles are needed
              return buildDefaultDragHandles
                  ? KeyedSubtree(
                      key: ValueKey('ReorderableItem${index}Key'),
                      child: Row(
                        mainAxisSize: MainAxisSize.min,
                        children: [
                          Expanded(child: item),
                          buildDefaultDragHandles
                              ? ReorderableDragStartListener(
                                  index: index,
                                  child: const Padding(
                                    padding: EdgeInsets.only(right: 8.0),
                                    child: Icon(Icons.drag_handle),
                                  ),
                                )
                              : const SizedBox.shrink(),
                        ],
                      ))
                  : item;
            }
          },
          onReorderStart: (index) {
            // Find the Separator widgets and hide them
            // Call onReorderStarted callback when the reorder starts
            onReorderStarted?.call();
          },
          onReorder: (int oldIndex, int newIndex) {
            if (oldIndex < newIndex) {
              newIndex -= 1;
            }

            if (oldIndex % 2 == 1) {
              //separator - should never happen
              return;
            }

            if ((oldIndex - newIndex).abs() == 1) {
              //moved behind the top/bottom separator
              return;
            }

            newIndex = oldIndex > newIndex && newIndex % 2 == 1
                ? (newIndex + 1) ~/ 2
                : newIndex ~/ 2;
            oldIndex = oldIndex ~/ 2;
            onReorder.call(oldIndex, newIndex);
          },
          onReorderEnd: (index) {
            onReorderEnded?.call(index);
          },
          itemExtent: itemExtent,
          prototypeItem: prototypeItem,
          proxyDecorator: proxyDecorator,
          buildDefaultDragHandles: false,
          padding: padding,
          header: header,
          scrollDirection: scrollDirection,
          reverse: reverse,
          scrollController: scrollController,
          primary: primary,
          physics: physics,
          shrinkWrap: shrinkWrap,
          anchor: anchor,
          cacheExtent: cacheExtent,
          dragStartBehavior: dragStartBehavior,
          keyboardDismissBehavior: keyboardDismissBehavior,
          restorationId: restorationId,
          clipBehavior: clipBehavior,
        );
}
```

Example:

```
class _MyHomePageState extends State<MyHomePage> with TickerProviderStateMixin {
  late AnimationController _animationController;

  @override
  void initState() {
    super.initState();
    _animationController = AnimationController(
      vsync: this,
      duration: const Duration(milliseconds: 300),
    );
  }

  final List<Widget> _list = List.generate(
      10,
      (index) => Container(
          key: ValueKey('abcd-$index'),
          width: 250,
          height: 150,
          color: index % 2 == 1 ? Colors.lightBlue : Colors.red,
          margin: const EdgeInsets.symmetric(vertical: 20, horizontal: 25)));
  final _scrollController = ScrollController();

  @override
  void dispose() {
    _animationController.dispose();
    super.dispose();
  }

  void _handleReorderStarted() {
    _animationController.forward();
  }

  void _handleReorderEnded(_) {
    Future.delayed(const Duration(milliseconds: 50), () {
      _animationController.reverse();
    });
  }

  @override
  Widget build(BuildContext context) {
    return Scaffold(
      appBar: AppBar(
        backgroundColor: Theme.of(context).colorScheme.inversePrimary,
        title: Text(widget.title),
      ),
      body: Center(
        child: CustomReorderableListView.separated(
          itemCount: _list.length,
          separatorBuilder: (context, index) => FadeTransition(
              opacity: _animationController.drive(Tween(begin: 1.0, end: 0.0)),
              child: Divider(key: ValueKey('separator-$index'))),
          itemBuilder: (_, int index) => _list[index],
          shrinkWrap: true,
          onReorder: (a, b) {
            final Widget item = _list.removeAt(a);
            _list.insert(b, item);
          },
          onReorderStarted: _handleReorderStarted,
          onReorderEnded: _handleReorderEnded,
          scrollController: _scrollController,
          proxyDecorator: (Widget child, _, animation) {
            return AnimatedBuilder(
              animation: animation,
              builder: (BuildContext context, Widget? child) {
                final animValue = Curves.easeInOut.transform(animation.value);
                final scale = lerpDouble(1, 1.05, animValue)!;
                final elevation = lerpDouble(0, 6, animValue)!;
                return Transform.scale(
                  scale: scale,
                  child: Material(
                    elevation: elevation,
                    borderRadius: BorderRadius.circular(0.1),
                    color: Colors.transparent,
                    child: child,
                  ),
                );
              },
              child: child,
            );
          },
        ),
      ),
    );
  }
}
```

**Comment by psaul20** [comment]
Any update here? I also would love to see a .separated constructor solution here!

**Comment by fanjin-z** [comment]
Any updates? I think this is a quite useful enhancement, I'd be happy to create a PR if nobody is currently working on it

**Comment by LucasTKP** [comment]
I removed the background colors from the element and put a bottom border to show the user which element they are moving:


ReorderableListView.builder(
              itemCount: controller.rotas.length,
              onReorder: controller.reorderRotas,
              proxyDecorator: (child, index, animation) => Material(
                  borderRadius: BorderRadius.circular(12),
                  color: Colors.transparent,
                  shadowColor: Colors.transparent,
                  borderOnForeground: false,
                  child: Container(
                    decoration: BoxDecoration(
                      borderRadius: BorderRadius.circular(12), 
                      border: const Border(
                        bottom: BorderSide(
                          color: Colors.grey,
                          width: 5.0,
                        ),
                      ),
                    ),
                    child: child,
                  )),
              itemBuilder: (context, index) {
                return KeyedSubtree(
                  key: ValueKey(controller.rotas[index].rotaDto.id),
                  child: Container(
                    margin: const EdgeInsets.only(top: 16),
                    child: CardRota(rotaModel: controller.rotas[index]),
                  ),
                );
              },
            ),

**Comment by krolmic** [comment]
Hi! I'd like to take this as my first contribution. Thanks for your input @victorsanni , @dkwingsmt , @darrenaustin and @Piinks !

### Proposed API

Mirror `ListView.separated` at the surface:

```dart
ReorderableListView.separated({
  required IndexedWidgetBuilder itemBuilder,
  required IndexedWidgetBuilder separatorBuilder,
  required int itemCount,
  ReorderItemCallback? onReorderItem,
  // ... 
```

- `itemExtent`, `itemExtentBuilder`, and `prototypeItem` are forced to `null`, items and separators may have different extents along the scroll axis.
- `onReorderItem` receives user-visible indices in `[0, itemCount]`, matching `ListView.separated`.

### Proposed internals (addressing the drag-alignment bug)

Rather than the `2n-1` wrapper pattern from the prior PR:

1. Add an internal `separatorBuilder` parameter to `SliverReorderableList`.
2. Its delegate produces `2n-1` children, but only the `n` item children are registered as reorderable `_ReorderableItem` slots. `_dragIndex`, `_insertIndex`, and all drag math stay at user-item granularity - no translation.
3. During drag, the sliver collapses exactly one adjacent separator:
   - trailing separator by default, or
   - leading separator when dragging the last item.

4. Separator assumptions stay inside the delegate, nothing leaks into semantics (addressing @Piinks's earlier feedback).

### Tests

New cases in `packages/flutter/test/material/reorderable_list_test.dart` ...

### Example

A `{@tool dartpad}` sample at `examples/api/lib/material/reorderable_list/reorderable_list_view.separated.0.dart`, with a smoke test under `examples/api/test/material/reorderable_list/`.

---

Could someone check this approach before I start implementing? Happy to open a PR soon.

**Comment by Piinks** [comment]
Hey @krolmic! Thanks for reaching out! That sounds like a good approach to me, very similar to https://github.com/flutter/flutter/pull/168834. I would take a look at that PR, and see what feedback was left unaddressed to ensure the same gaps are not present. Thanks very much, I'll be happy to review when you have something to take a look at! :)

**Comment by krolmic** [comment]
> Hey [@krolmic](https://github.com/krolmic)! Thanks for reaching out! That sounds like a good approach to me, very similar to [#168834](https://github.com/flutter/flutter/pull/168834). I would take a look at that PR, and see what feedback was left unaddressed to ensure the same gaps are not present. Thanks very much, I'll be happy to review when you have something to take a look at! :)

Cool! 🤩 Thanks for your confirmation @Piinks . I'm going to start working on it today and open a PR soon.

**Comment by krolmic** [comment]
Hi @Piinks! The PR is open: #190131. It looks like CI is waiting on a maintainer approval to run. Happy to hear your thoughts whenever you have time!
